larpandoracontent/LArTwoDReco/TrackInEMShowerAlgorithm.cc

```diff
@@ -0,0 +1,892 @@
+    
+    ClusterVector clusterVector;
+    this->SelectCleanClusters(pClusterList, clusterVector);
+
+    TwoDSlidingFitResultMap microSlidingFitResultMap;
+    TwoDSlidingFitResultMap macroSlidingFitResultMap;
+    this->InitialiseSlidingFitResultMaps(clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
+
+    // Find longest length association and conitnue merging until no further merges can be made
+    bool mergeMade(false);
+    do
+    {
+        ClusterAssociation clusterAssociation;
+        if(!this->FindBestClusterAssociation(clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap, clusterAssociation))
+            break;
+        
+        ClusterToCaloHitListMap clusterToCaloHitListMap;
+        CaloHitVector extrapolatedCaloHitVector;
+        this->GetExtrapolatedCaloHits(clusterAssociation, pClusterList, extrapolatedCaloHitVector, clusterToCaloHitListMap);
+        
+        if (extrapolatedCaloHitVector.empty())
+            break;
+        
+        if (!this->IsTrackContinuous(clusterAssociation, extrapolatedCaloHitVector))
+            break;
+        
+        this->RefineTracks(clusterAssociation, microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector, extrapolatedCaloHitVector);
+        
+        this->MergeHits(clusterAssociation, clusterToCaloHitListMap, microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector);
+
+        this->UpdateSlidingFitResultMap(clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
+
+        mergeMade = true;
+    } while (mergeMade);
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackInEMShowerAlgorithm::SelectCleanClusters(const ClusterList *pClusterList, ClusterVector &clusterVector) const
+{
+    for (const Cluster *const pCluster : *pClusterList)
+    {
+        if (pCluster->GetNCaloHits() < m_minCaloHits)
+            continue;
+
+        clusterVector.push_back(pCluster);
+    }
+    
+    std::sort(clusterVector.begin(), clusterVector.end(), LArClusterHelper::SortByNHits);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackInEMShowerAlgorithm::InitialiseSlidingFitResultMaps(const ClusterVector &clusterVector, TwoDSlidingFitResultMap &microSlidingFitResultMap,
+    TwoDSlidingFitResultMap &macroSlidingFitResultMap) const
+{
+    const float slidingFitPitch(LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+    const unsigned int macroSlidingFitWindow(1000);
+    
+    for (const Cluster *const pCluster : clusterVector)
+    {
+        try
+        {
+            (void) microSlidingFitResultMap.insert(TwoDSlidingFitResultMap::value_type(pCluster, TwoDSlidingFitResult(pCluster, m_slidingFitWindow, slidingFitPitch)));
```

Comment:
This has been on my mind. 

The clusters contained within a cluster association are guaranteed to have an entry in both maps - this is because when forming the cluster association object a cluster pair if skipped if an entry is not found in both maps. Only the fits of the clusters in the cluster association are ever accessed. 

But I think that i'll demand this to be the case here so the map is smaller

Reply:
Great, the main aim of changes such as this is to be sure that things will behave in those exceptional cases where things aren't the standard, expected use case. If we think about what might go wrong each step, then hopefully we'll pre-empt most problems and produce something nice and robust.